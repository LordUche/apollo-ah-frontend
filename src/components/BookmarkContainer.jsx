/* eslint-disable react/jsx-indent */
/* eslint-disable react/jsx-closing-tag-location */
/* eslint-disable react/no-array-index-key */
/* eslint-disable react/jsx-wrap-multilines */
import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import BookmarkItem from '../views/BookmarkItem';
import { getBookmarkedArticles } from '../actions/getBookmark';
import * as selectors from '../selectors/getBookmarkSelector';

/**
 * @description - returns all articles bookmarked by a user
 * @return {JSX}
 */
function BookmarkContainer(props) {
  const [viewType, setViewType] = useState('row view-row');

  const { articles } = props;

  const getArticlesHandler = () => {
    props.getBookmarkedArticles();
  };

  const changeViewToHorizontal = () => {
    setViewType('row view-row list-view');
  };

  const changeViewToVertical = () => {
    setViewType('row view-row');
  };

  useEffect(() => (
    getArticlesHandler()
  ), []);

  return (
    <>
      <header className="page-header" data-pg-collapsed>
        <h1>Bookmarks </h1>
        <div className="btn-group btn-group-toggle float-right view-changer" data-toggle="buttons">
          <label className="btn active" onClick={changeViewToVertical}>
            <input type="radio" name="options" autoComplete="off" checked value="" />
            <i className="fas fa-th" />
          </label>
          <label className="btn" onClick={changeViewToHorizontal}>
            <input type="radio" name="options" autoComplete="off" value="list-view" />
            <i className="fas fa-list" />
          </label>
        </div>
      </header>
      {
        articles.articles.length !== 0
          ? (
            <>
              <div className={viewType}>
                {
                  // eslint-disable-next-line arrow-body-style
                  articles.articles.map((article, indx) => {
                    return (
                      <BookmarkItem
                        key={indx}
                        category={article.Article.articleCategory.category}
                        image={article.Article.image}
                        articleTitle={article.Article.title}
                        articleDescription={article.Article.description}
                        authorsName={article.Article.User.Profile.firstname}
                        authorsImage={article.Article.User.Profile.image}
                        readTime={article.Article.readTime}
                      />
                    );
                  })
                }
              </div>
              {/** pagination support to be implemented later */}
              <div>Pagination is here</div>
            </>
          ) : <div>
            <i className="fas fa-book-open" style={{ fontSize: '70px', color: '#66008c' }} />
            <p style={{ paddingTop: '30px' }}>*It looks like you have not bookmarked any article yet</p>
          </div>
      }
    </>
  );
}

const mapStateToProps = createStructuredSelector(
  {
    isLoading: selectors.getBookmarkLoading,
    articles: selectors.getBookmarkArticles,
  }
);

export default connect(mapStateToProps, { getBookmarkedArticles })(BookmarkContainer);
