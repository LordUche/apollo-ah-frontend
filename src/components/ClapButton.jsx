/* eslint-disable no-undef */
import React, { Component } from 'react';
import PropType from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { debounce } from 'lodash';
import { withRouter } from 'react-router-dom';
import clapIcon from '../images/clap.svg';
import * as articleSelector from '../selectors/singleArticleSelector';
import { getAuthId, getIsAuthenticated } from '../selectors/authSelector';
import { getToken } from '../selectors/profileSelector';
import { clapArticleRequest, fetchUserClaps } from '../actions/clapsAction';
import { getUserClaps } from '../selectors/clapsSelector';
import { CLAPS_LIMIT, CLAPS_REQUEST_TIMER } from '../utils/contants';

/**
 * A class representing clap button component
 *
 * @class ClapButton
 * @extends {Component}
 */
export class ClapButton extends Component {
  /**
   * Handles clap request (debounced)
   *
   * @returns {void}
   * @memberof ClapButton
   */
  makeClapRequest = debounce(() => {
    const { clapArticle, articleSlug: slug, token } = this.props;
    const { clapsCount: claps } = this.state;
    this.setState({ clapsCount: 0 });
    return clapArticle({ slug, claps, token });
  }, CLAPS_REQUEST_TIMER);

  static propTypes = {
    articleSlug: PropType.string.isRequired,
    articleAuthorId: PropType.number.isRequired,
    articleClaps: PropType.number.isRequired,
    clapArticle: PropType.func.isRequired,
    isLoggedIn: PropType.bool.isRequired,
    userId: PropType.number,
    history: PropType.object.isRequired,
    token: PropType.string.isRequired,
    userClaps: PropType.number.isRequired,
    loadUserClaps: PropType.func.isRequired,
  };

  static defaultProps = {
    userId: null,
  };

  constructor(props) {
    super(props);
    this.state = {
      clapsLeft: CLAPS_LIMIT - props.userClaps,
      articleClaps: props.articleClaps,
      clapsCount: 0, // record clap button clicks
    };
  }

  componentDidMount = async () => {
    const {
      articleSlug: slug,
      isLoggedIn,
      userId,
      loadUserClaps
    } = this.props;
    if (isLoggedIn) {
      loadUserClaps({ slug, userId });
    }
  }

  /**
   * Makes an article clap
   *
   * @returns {void}
   * @memberof ClapButton
   */
  handleClapArticle = () => {
    const {
      isLoggedIn,
      userId,
      articleAuthorId,
      history,
    } = this.props;

    if (!isLoggedIn) { // redirect guest users to login
      history.push('/login');
      return false;
    }
    // Deny article's author
    if (articleAuthorId === userId) return false;
    const { articleClaps, clapsLeft, clapsCount } = this.state;
    if (clapsLeft >= 1) {
      this.setState({
        clapsLeft: clapsLeft - 1,
        articleClaps: articleClaps + 1,
        clapsCount: clapsCount + 1,
      });
      this.makeClapRequest();
    }
  }

  /**
   * React render function
   *
   * @returns {JSX.Element} DOM element
   * @memberof ClapButton
   */
  render() {
    const { articleClaps } = this.state;

    return (
      <span className="clap-icon">
        <button type="button" className="btn" onClick={this.handleClapArticle}>
          <img src={clapIcon} alt="" />
        </button>
        {articleClaps}
        {' '}
        claps
      </span>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  articleSlug: articleSelector.getArticleSlug,
  articleAuthorId: articleSelector.getArticleAuthorId,
  articleClaps: articleSelector.getArticleClaps,
  isLoggedIn: getIsAuthenticated,
  userId: getAuthId,
  token: getToken,
  userClaps: getUserClaps,
});

const mapDispatchToProps = {
  clapArticle: claps => clapArticleRequest(claps),
  loadUserClaps: payload => fetchUserClaps(payload),
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(ClapButton));
