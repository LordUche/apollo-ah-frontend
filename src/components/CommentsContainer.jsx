/* eslint-disable react/prefer-stateless-function */
import React, { Component } from 'react';

import Comments from '../views/Comments';

/**
 * Container component for the Article comments view
 * @export
 * @class CommentsContainer
 * @extends {Component}
 */
export default class CommentsContainer extends Component {
  render() {
    return (
      <Comments />
    );
  }
}
