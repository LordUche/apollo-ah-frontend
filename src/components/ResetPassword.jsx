/* eslint-disable import/no-named-as-default */
/* eslint-disable no-unused-expressions */
import React, { Component } from 'react';
import jwtDecode from 'jwt-decode';
import PropTypes from 'prop-types';
import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';
import * as selectors from '../selectors/resetPassword';
import { resetPassword } from '../actions/resetPasswordAction';
import ResetPasswordForm from '../views/ResetPasswordForm';

/**
 * Form for resetting password
 * @class UpdateUserPassword
 * @extends {Component}
 * @param {object} props Component props
 * @param {function} props.updatePassword action creator function
 * @returns {JSX}
 */
export class ResetPassword extends Component {
  // local state for handling email gotten from the input field
  state = {
    password: null,
    confirmPassword: null,
  }

  // this is a function to update the password and confirmPassword state
  updateInput = (event) => {
    this.setState({
      [event.target.name]: event.target.value
    });
  };

  // A function that triggers the post request action
  handleSubmit = async (event) => {
    event.preventDefault();
    const token = new URLSearchParams(window.location.search).get('token');
    const { email } = jwtDecode(token);
    const { password, confirmPassword } = this.state;
    const validateInput = /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{6,}$/;
    if (!password.match(validateInput)) {
      toast.error('Password must contain uppercase and lowercase letters, a number and not be less than 6 characters');
    } else if (password === confirmPassword) {
      // calling the action creator to update the store
      const { resetPassword, history } = this.props;
      await resetPassword(token, password, confirmPassword, email);
      const { message } = this.props;
      message === 'Password reset successful'
        ? toast.success('Password updated successfully !')
        : toast.error('Your request to reset password has expired or check your internet connection.');
      history.push('/login');
    } else {
      toast.error('Password do not match');
    }
  };

  /**
   * @memberof ResetPassword
   * @returns {JSX} The JSX representation of the Reset Password component
  */
  render() {
    return (
      <ResetPasswordForm
        handleSubmit={this.handleSubmit}
        updateInput={this.updateInput}
      />
    );
  }
}

ResetPassword.propTypes = {
  resetPassword: PropTypes.func,
  history: PropTypes.object,
  message: PropTypes.string,
};

ResetPassword.defaultProps = {
  resetPassword: () => {},
};

const mapStateToProps = createStructuredSelector({
  message: selectors.getResetPasswordMessage,
});


export default connect(mapStateToProps, { resetPassword })(ResetPassword);
