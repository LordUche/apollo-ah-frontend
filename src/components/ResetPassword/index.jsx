/* eslint-disable react/prop-types */
import React, { useState, useRef } from 'react';
import { connect } from 'react-redux';
import { ClipLoader } from 'react-spinners';
import Modal from '../../views/Modal.js';
import passwordResetRequest from '../../actions/resetPassword';


/**
 * Form for resetting password
 * @param {object} props Component props
 * @param {function} props.passwordResetRequest action creator function
 */
function ResetPassword(props) {
  const [userEmail, setUserEmail] = useState({
    email: null,
  });

  const { isLoading } = props;

  const [message, setMessage] = useState(null);
  const input = useRef(null);
  
  const updateInput = event => {
    setUserEmail({ ...userEmail, [event.target.name]: event.target.value });
  };

  const clearInputs = () => {
    input.current.value = null;
    setUserEmail({ ...userEmail, email: null });
  };

  const handleSubmit = () => {
    setMessage(null);
    props.passwordResetRequest(userEmail.email).then(res => {
      clearInputs();
      if (res.code === 400) {
        setMessage(res.data[0].message);
      }
      if (res.code === 200) {
        setMessage(res.message);
      }
    });
  };

  const validateUserData = (event) => {
    event.preventDefault();
    const fieldsArr = Object.entries(userEmail);
    const errors = [];
    fieldsArr.forEach(field => {
      if (!field[1]) {
        errors.push(`${field[0]} is required.`);
      }
    });
    if (errors.length < 1) {
      return handleSubmit(event);
    } else {
      setMessage(errors);
    }
  };

  return (
    <Modal clearMessage={() => setMessage(null)} title="Recover Password!">
      <p className="reset_password_paragraph">Enter the email address you used when you created your account. You will receive instructions to reset your password in your mail box.</p>
      <form>
        <div className="form-group">
          <input ref={input} type="email" className="reset_form_control" id="email" placeholder="Enter email" name="email" onChange={(event) => updateInput(event)} />
        </div>
        <button type="submit" className="reset_btn" onClick={(event => validateUserData(event))}>Request reset link</button>
      </form>
      <div>
        {
          isLoading ?
            (
              <div className="reset_spinner_box">
                <ClipLoader
                  size={40}
                  color="purple"
                />
              </div>
            ) : null
        }
      </div>
      {
        message ? <div className="reset_alert_danger">{message}</div> : null
      }
    </Modal>
  );
}

const mapStateToProps = state => ({
  isLoading: state.resetPasswordReducer.isLoading,
});

export default connect(mapStateToProps, { passwordResetRequest })(ResetPassword);
