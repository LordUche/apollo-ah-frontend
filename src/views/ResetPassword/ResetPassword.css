.ModalContent {
  position: relative;
  display: -ms-flexbox;
  display: flex;
  width: 661px;
  height: 445px;
  -ms-flex-direction: column;
  flex-direction: column;
  pointer-events: auto;
  background-color: #fff;
  background-clip: padding-box;
  border: 1px solid rgba(0,0,0,.2);
  border-radius: .3rem;
  outline: 0;
}

.ModalHeader {
  display: -ms-flexbox;
  display: flex;
  -ms-flex-align: start;
  align-items: flex-start;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 1rem 1rem;
  border-top-left-radius: .3rem;
  border-top-right-radius: .3rem;
}

.ModalTitle {
  margin-bottom: 0;
  line-height: 1.5;
  font-size: 30px;
  font-weight: bold;
}

.ModalBody {
  position: relative;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  padding: 1rem 4rem;
}

.Close {
  padding: 0;
  background-color: transparent;
  border: 0;
  -webkit-appearance: none;
  padding-right: 25px;
  margin-top: 17px;
  font-size: 31px;
  -moz-appearance: none;
  appearance: none;
  float: right;
  font-size: 1.5rem;
  font-weight: 700;
  line-height: 1;
  color: #000;
  text-shadow: 0 1px 0 #fff;
  opacity: .5;
}

.Paragraph {
  margin-left: 16px;
  font-size: 13px;
  color: #968d8d;
}

.FormControl {
  overflow: visible;
  display: block;
  width: 94%;
  margin-left: 14px;
  height: calc(1.5em + .75rem + 21px);
  padding: .375rem .75rem;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #495057;
  outline: none;
  background-color: #f3f3f3;
  background-clip: padding-box;
  border: 1px solid #ced4da;
  border-radius: .25rem;
  transition: border-color .15s ease-in-out,box-shadow .15s ease-in-out;
}

.Btn {
  display: inline-block;
  font-weight: 400;
  color: #212529;
  text-align: center;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #fff;
  background-color: #65018c;
  border-color: #007bff;
  margin-left: 12px;
  height: calc(1.5em + .75rem + 21px);
  width: 501px;
  border: 1px solid transparent;
  padding: .375rem .75rem;
  font-size: 1rem;
  line-height: 1.5;
  border-radius: .25rem;
  outline: none;
  transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;
}