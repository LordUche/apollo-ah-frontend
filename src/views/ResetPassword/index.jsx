import React, { useState } from 'react';
import { connect } from 'react-redux';
import { passwordResetRequest } from '../../actions/resetPassword';
import classes from './ResetPassword.css';

function ResetPassword(props) {
  const [userEmail, setUserEmail] = useState({
    email: null,
  });
  
  const updateInput = event => {
    setUserEmail({ ...userEmail, [event.target.name]: event.target.value });
  };

  const handleSubmit = () => {
    props.passwordResetRequest(userEmail.email);
  };

  const validateUserData = (event) => {
    event.preventDefault();
    // setLoading(true);
    const fieldsArr = Object.entries(userEmail);
    const errors = [];
    fieldsArr.forEach(field => {
      if (!field[1]) {
        errors.push(`${field[0]} is required.`);
        // props.failureToast(`${field[0]} is required.`);
      }
    });
    return errors.length < 1 ? handleSubmit() : console.log('--->', errors);
  };

  return (
    <div className="modal" id="myModal">
      <div className="modal-dialog">
        <div className={classes.ModalContent}>
          <div className={classes.cancelTop}>
            <button type="button" className={classes.Close} data-dismiss="modal">&times;</button>
          </div>
          <div className={classes.ModalBody}>
            <div className={classes.ModalHeader}>
              <h4 className={classes.ModalTitle}>Recover Password!</h4>
            </div>
            <p className={classes.Paragraph}>Enter the email address you used when you created your account. You will receive instructions to reset your password in your mail box.</p>
            <form>
              <div className="form-group">
                <input type="email" className={classes.FormControl} id="email" placeholder="Enter email" name="email" onChange={(event) => updateInput(event)} />
              </div>
              <button type="submit" className={classes.Btn} onClick={(event => validateUserData(event))}>Request reset link</button>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
}

// export default ResetPassword;
export default connect(() => ({}), { passwordResetRequest })(ResetPassword);
