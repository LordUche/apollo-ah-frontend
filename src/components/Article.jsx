import React from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import ArticleBody from '../views/ArticleBody';
import CommentsContainer from './CommentsContainer';

/**
 *
 *
 * @class Article
 * @extends {React.Component}
 */
class Article extends React.Component {
  /**
   *
   * @static - Validating proptypes
   * @memberof Article
   */
  static propTypes = {
    match: PropTypes.object.isRequired,
  }

  /**
   * Creates an instance of Article.
   * @param {*} props - Props passed down from the BrowserRouter
   * @memberof Article
   */
  constructor(props) {
    super(props);
    this.state = {
      article: {},
      bookmarked: false,
    };
  }

  /**
   *
   * @todo - Move the fetch article functionality to redux store
   * @memberof Article
   */
  componentDidMount = async () => {
    const { match } = this.props;
    const { slug } = match.params;
    const result = await axios.get(`${process.env.API_BASE_URL}/articles/${slug}`);
    this.setState({ article: result.data.data });
  }

  /**
   * @todo - employ javascript debounce to avoid calling API on repeated clicks
   * @method bookmarkArticle - The bookmark article functionality
   * @memberof Article
   */
  bookmarkArticle = () => {
    const { bookmarked } = this.state;
    this.setState({ bookmarked: !bookmarked });
  };

  /**
   *
   *
   * @returns {JSX} The JSX representation of the Article component
   * @memberof Article
   */
  render() {
    const { article, bookmarked } = this.state;
    return (
      <div>
        <main className="main-body">
          <section className="min-vh-100">
            {article.image && (
              <div
                className="pg-empty-placeholder single-hero"
                style={{ backgroundImage: `url(${article.image})` }}
              />
            )}
            <div className="container">
              <div className="single-container">
                <ArticleBody
                  article={article}
                  bookmarkArticle={this.bookmarkArticle}
                  bookmarked={bookmarked}
                />
                <CommentsContainer />
              </div>
            </div>
            <div className="single-suggested-grp">
              <div className="container">
                <h4>Recommended for you</h4>
                <div className="row pg-empty-placeholder" />
              </div>
            </div>
          </section>
        </main>
      </div>
    );
  }
}

export default Article;
