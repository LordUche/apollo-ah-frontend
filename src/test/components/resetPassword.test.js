import React from 'react';
import { shallow } from 'enzyme';
import { ResetPassword } from '../../components/ResetPassword';

describe('<ResetPassword Test Suite>', () => {
  const props = {
    isLoading: false,
    resetPassword: jest.fn(),
    message: 'Password updated successfully',
    history: {
      push: jest.fn()
    }
  };
  const event = {
    target: {
      name: 'password',
      value: 'mypassword'
    },
    preventDefault: jest.fn(),
  };
  it('It should render succesfully when user is not logged in', () => {
    const wrapper = shallow(<ResetPassword {...props} />);
    expect(wrapper).toBeDefined();
    expect(wrapper.length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });
  it('should update password with users password', () => {
    const wrapper = shallow(<ResetPassword {...props} />);
    wrapper.instance().updateInput(event);
    expect(wrapper.instance().state.password).toEqual('mypassword');
  });
  it('should change users password', async () => {
    const wrapper = shallow(<ResetPassword {...props} />);
    await wrapper.setState({
      password: 'Password1',
      confirmPassword: 'Password1'
    });
    wrapper.instance().handleSubmit(event);
  });
});
