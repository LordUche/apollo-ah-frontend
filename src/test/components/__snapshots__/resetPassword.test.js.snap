// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetPassword Test Suite> It should render succesfully when user is not logged in 1`] = `
<ResetPasswordForm
  handleSubmit={[Function]}
  updateInput={[Function]}
/>
`;
